src/ddsrt/src/events/darwin/events_darwin.c

```diff
@@ -0,0 +1,236 @@
+*/
+static dds_return_t ddsrt_event_queue_init(ddsrt_event_queue_t* queue) {
+  queue->nevents = 0;
+  queue->cevents = 8;
+  queue->ievents = 0;
+  ddsrt_mutex_init(&queue->lock);
+
+  /*create kevent polling instance */
+  queue->kq = kqueue();
+  assert(queue->kq != -1);
+  assert(fcntl(queue->kq, F_SETFD, fcntl(queue->kq, F_GETFD) | FD_CLOEXEC) != -1);
+
+  queue->events = ddsrt_malloc(sizeof(ddsrt_event_t*) * queue->cevents);
+  assert(queue->events);
+
+#if !defined(LWIP_SOCKET)
+  if (-1 == pipe(queue->interrupt))
+    return DDS_RETCODE_ERROR;
+  /*register interrupt event*/
+  struct kevent kev;
+  EV_SET(&kev, queue->interrupt[0], EVFILT_READ, EV_ADD, 0, 0, 0);
+  assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+#endif /* !LWIP_SOCKET */
+
+  /*create kevents array*/
+  queue->kevents = ddsrt_malloc(sizeof(struct kevent) * queue->cevents);
+  assert(queue->kevents);
+
+  return DDS_RETCODE_OK;
+}
+
+/**
+* @brief Finishes an event queue.
+*
+* Will free created containers and do any additional cleanup of things created in ddsrt_event_queue_init.
+*
+* @param[out] queue The queue to finish.
+*
+* @returns DDS_RETCODE_OK if everything went OK.
+*/
+static dds_return_t ddsrt_event_queue_fini(ddsrt_event_queue_t* queue) {
+  assert(queue);
+  /*deregister all currently stored triggers from kqueue*/
+  struct kevent kev;
+  for (unsigned int i = 0; i < queue->nevents; i++) {
+    ddsrt_event_t* evt = queue->events[i];
+    if (evt->flags & DDSRT_EVENT_FLAG_READ)
+      EV_SET(&kev, evt->data.socket.sock, EVFILT_READ, EV_DELETE, 0, 0, evt);
+    assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+  }
+#if !defined(LWIP_SOCKET)
+  EV_SET(&kev, queue->interrupt[0], EVFILT_READ, EV_DELETE, 0, 0, 0);
+  assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+  close(queue->interrupt[0]);
+  close(queue->interrupt[1]);
+#endif /* !LWIP_SOCKET */
+  close(queue->kq);
+
+  ddsrt_mutex_destroy(&queue->lock);
+  ddsrt_free(queue->events);
+  ddsrt_free(queue->kevents);
+  return DDS_RETCODE_OK;
+}
+
+ddsrt_event_queue_t* ddsrt_event_queue_create(void) {
+  ddsrt_event_queue_t* returnptr = ddsrt_malloc(sizeof(ddsrt_event_queue_t));
+  assert(returnptr);
+  ddsrt_event_queue_init(returnptr);
+  return returnptr;
+}
+
+dds_return_t ddsrt_event_queue_destroy(ddsrt_event_queue_t* queue) {
+  assert(queue);
+  ddsrt_event_queue_fini(queue);
+  ddsrt_free(queue);
+  return DDS_RETCODE_OK;
+}
+
+size_t ddsrt_event_queue_nevents(ddsrt_event_queue_t* queue) {
+  size_t ret;
+  ddsrt_mutex_lock(&queue->lock);
+  ret = queue->nevents;
+  ddsrt_mutex_unlock(&queue->lock);
+  return ret;
+}
+
+dds_return_t ddsrt_event_queue_wait(ddsrt_event_queue_t* queue, dds_duration_t reltime) {
+  assert(queue);
+
+  /*reset triggered status*/
+  ddsrt_mutex_lock(&queue->lock);
+  queue->ievents = 0;
+  for (unsigned int i = 0; i < queue->nevents; queue++)
+    ddsrt_atomic_st32(&queue->events[i]->triggered, DDSRT_EVENT_FLAG_UNSET);
```

Comment:
I'm not too happy about writing to all events (which are scattered in memory). It is probably unavoidable short of embedding the events in the `ddsrt_event_queue_t`, rather than letting the latter point to events owned elsewhere.

(Not that one can always avoid all cache misses ... but this seems like a place where making an effort to significantly reduce the number of them would pay off.)

Reply:
The events do not have to be scattered in memory, if the entity using the event queue just has the events which are used in the queue has them in some kind of flat array, and just passes pointers to those entries to the queue.